fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SubqueryTransformer.java

```diff
@@ -45,95 +48,102 @@ public SubqueryTransformer(ConnectContext session) {
 
     public OptExprBuilder handleSubqueries(ColumnRefFactory columnRefFactory, OptExprBuilder subOpt, Expr expression,
                                            CTETransformerContext cteContext) {
-        if (subOpt.getExpressionMapping().hasExpression(expression)) {
-            return subOpt;
-        }
-
-        FilterWithSubqueryHandler handler = new FilterWithSubqueryHandler(columnRefFactory, session);
-        subOpt = expression.accept(handler, new SubqueryContext(subOpt, true, cteContext));
-
-        return subOpt;
+        return doHandleSubqueries(columnRefFactory, subOpt, expression, cteContext, true);
     }
 
-    // Only support scalar-subquery in `SELECT` clause
     public OptExprBuilder handleScalarSubqueries(ColumnRefFactory columnRefFactory, OptExprBuilder subOpt,
                                                  Expr expression, CTETransformerContext cteContext) {
+        return doHandleSubqueries(columnRefFactory, subOpt, expression, cteContext, false);
+    }
+
+    private OptExprBuilder doHandleSubqueries(ColumnRefFactory columnRefFactory, OptExprBuilder subOpt, Expr expression,
+                                              CTETransformerContext cteContext, boolean useSemiAnti) {
         if (subOpt.getExpressionMapping().hasExpression(expression)) {
             return subOpt;
         }
 
         FilterWithSubqueryHandler handler = new FilterWithSubqueryHandler(columnRefFactory, session);
-        subOpt = expression.accept(handler, new SubqueryContext(subOpt, false, cteContext));
+        subOpt = expression.accept(handler, new SubqueryContext(subOpt, useSemiAnti, cteContext));
 
         return subOpt;
     }
 
-    public ScalarOperator rewriteSubqueryScalarOperator(Expr predicate, OptExprBuilder subOpt,
-                                                        List<ColumnRefOperator> correlation) {
+    public OptExprBuilder handlePredicate(OptExprBuilder subOpt, Expr predicate, List<ColumnRefOperator> correlation) {
         ScalarOperator scalarPredicate =
                 SqlToScalarOperatorTranslator.translate(predicate, subOpt.getExpressionMapping(), correlation);
 
+        subOpt = eliminateUnusedApplyNodesFrom(subOpt, scalarPredicate);
```

Comment:
Why not just use ScalarOperatorRewriter to rewrite the scalarPredicate?

Reply:
Thanks, good suggestion, and I add another rule apply mode `OneShot`, please review again